contracts/governance/ProposalValidator.sol

```diff
@@ -3,198 +3,253 @@ pragma solidity 0.7.6;
+   * @return boolean, true if can be cancelled
    **/
-  function getMinimumPropositionPowerNeeded(IKyberGovernance governance, uint256 blockNumber)
-    public
-    override
-    view
-    returns (uint256)
-  {
-    IGovernanceStrategy currentGovernanceStrategy = IGovernanceStrategy(
-      governance.getGovernanceStrategy()
+  function validateProposalCancellation(
+    IKyberGovernance governance,
+    uint256 proposalId,
+    address user
+  ) external view returns (bool) {
+    IKyberGovernance.ProposalWithoutVote memory proposal = governance.getProposalById(proposalId);
+    return (
+      strategy.getVotingPower(user, timestamp) >= 
+      proposal.maxVotingPower.mul(PROPOSITION_THRESHOLD).div(ONE_HUNDRED_WITH_PRECISION)
     );
-    return
-      currentGovernanceStrategy
-        .getTotalPropositionSupplyAt(blockNumber)
-        .mul(PROPOSITION_THRESHOLD)
-        .div(ONE_HUNDRED_WITH_PRECISION);
   }
 
   /**
-   * @dev Returns whether a proposal passed or not
-   * @param governance Governance Contract
+   * @dev Returns whether a binary proposal passed or not
+   * @param governance governance contract to fetch proposals from
    * @param proposalId Id of the proposal to set
    * @return true if proposal passed
    **/
-  function isProposalPassed(IKyberGovernance governance, uint256 proposalId)
-    external
-    override
+  function isBinaryProposalPassed(IKyberGovernance governance, uint256 proposalId)
+    public
     view
+    override
     returns (bool)
   {
     return (isQuorumValid(governance, proposalId) &&
       isVoteDifferentialValid(governance, proposalId));
   }
 
   /**
-   * @dev Calculates the minimum amount of Voting Power needed for a proposal to Pass
-   * @param votingSupply Total number of oustanding voting tokens
-   * @return voting power needed for a proposal to pass
+   * @dev Check whether a proposal has reached quorum
+   * @param governance governance contract to fetch proposals from
+   * @param proposalId Id of the proposal to verify
+   * @return true if minimum quorum is reached
    **/
-  function getMinimumVotingPowerNeeded(uint256 votingSupply)
+  function isQuorumValid(IKyberGovernance governance, uint256 proposalId)
     public
     override
     view
-    returns (uint256)
+    returns (bool)
   {
-    return votingSupply.mul(MINIMUM_QUORUM).div(ONE_HUNDRED_WITH_PRECISION);
+    IKyberGovernance.ProposalWithoutVote memory proposal = governance.getProposalById(proposalId);
+    // TODO: check quorum definition used is in line with what we want
+    if (proposal.proposalType == IKyberGovernance.ProposalType.Binary) {
+      return isMinimumQuorumReached(proposal.voteCounts[YES_INDEX], proposal.maxVotingPower);
+    } else if (proposal.proposalType == IKyberGovernance.ProposalType.Generic) {
+      (,uint256 winningOptionVoteCount) = getWinningOptionData(proposal.voteCounts);
+      return isMinimumQuorumReached(winningOptionVoteCount, proposal.maxVotingPower);
```

Comment:
is it correct for generic proposal or this func is only used for binary one?

Reply:
agree. `isVoteDifferentialValid ` is not support for generic proposal so this should not support for it

Reply:
`isVoteDifferentialValid` now supports both proposal types as well